File: ffi/ts/src/ctx.ts
import {
  addToTxInVec,
  addToTxOutVec,
  BLSCT_IN_AMOUNT_ERROR,
  BLSCT_OUT_AMOUNT_ERROR,
  buildCTx,
  castToUint8_tPtr,
  createTxInVec,
  createTxOutVec,
  deleteCTx,
  deleteTxInVec,
  deleteTxOutVec,
  freeObj,
  getCTxId,
  getCTxIns,
  getCTxOuts,
  hexToMallocedBuf,
  toHex,
} from './blsct'

import { CTxId } from './ctxId'
import { CTxIns } from './ctxIns'
import { CTxOuts } from './ctxOuts'
import { ManagedObj } from './managedObj'
import { TxIn } from './txIn'
import { TxOut } from './txOut'

/** Represents a confidential transaction. Also known as `CMutableTransaction` on the C++ side.
 * Examples:
 * ```ts
 * const { CTx, CTxId, TxIn, TxOut, ChildKey, SpendingKey, SubAddr, SubAddrId, DoublePublicKey, OutPoint, PublicKey, TokenId, CTX_ID_SIZE } = require('navio-blsct')
 * const { randomBytes } = require('crypto')
 * const cTxIdHex = randomBytes(CTX_ID_SIZE).toString('hex')
 * const cTxId = CTxId.deserialize(cTxIdHex)
 * const numTxIn = 1
 * const numTxOut = 1
 * const defaultFee = 200000
 * const fee = (numTxIn + numTxOut) * defaultFee
 * const outAmount = 10000
 * const inAmount = fee + outAmount
 * const outIndex = 0
 * const outPoint = OutPoint.generate(cTxId, outIndex)
 * const gamma = 100
 * const spendingKey = new SpendingKey()
 * const tokenId = TokenId.default()
 * const txIn = TxIn.generate(inAmount, gamma, spendingKey, tokenId, outPoint)
 * const viewKey = new ChildKey().toTxKey().toViewKey()
 * const spendingPubKey = new PublicKey()
 * const subAddrId = SubAddrId.generate(123, 456)
 * const subAddr = SubAddr.generate(viewKey, spendingPubKey, subAddrId)
 * const txOut = TxOut.generate(subAddr, outAmount, 'navio')
 * const cTx = CTx.generate([txIn], [txOut])
 * for (const cTxIn of cTx.getCTxIns()) {
 *   console.log(`prevOutHash: ${cTxIn.getPrevOutHash()}`)
 *   console.log(`prevOutN: ${cTxIn.getPrevOutN()}`)
 *   console.log(`scriptSig: ${cTxIn.getScriptSig()}`)
 *   console.log(`sequence: ${cTxIn.getSequence()}`)
 *   console.log(`scriptWitness: ${cTxIn.getScriptWitness()}`)
 * }
 * for (const cTxOut of cTx.getCTxOuts()) {
 *   console.log(`value: ${cTxOut.getValue()}`)
 *   console.log(`script_pub_key: ${cTxOut.getScriptPubKey()}`)
 *   console.log(`blsctData.spendingKey: ${cTxOut.blsctData().getSpendingKey()}`)
 *   console.log(`blsctData.ephemeralKey: ${cTxOut.blsctData().getEphemeralKey()}`)
 *   console.log(`blsctData.blindingKey: ${cTxOut.blsctData().getBlindingKey()}`)
 *   console.log(`blsctData.viewTag: ${cTxOut.blsctData().getViewTag()}`)
 *   const rp = cTxOut.blsctData().getRangeProof()
 *   console.log(`blsctData.rangeProof.A: ${rp.get_A()}`)
 *   console.log(`blsctData.rangeProof.A_wip: ${rp.get_A_wip()}`)
 *   console.log(`blsctData.rangeProof.B: ${rp.get_B()}`)
 *   console.log(`blsctData.rangeProof.r_prime: ${rp.get_r_prime()}`)
 *   console.log(`blsctData.rangeProof.s_prime: ${rp.get_s_prime()}`)
 *   console.log(`blsctData.rangeProof.delta_prime: ${rp.get_delta_prime()}`)
 *   console.log(`blsctData.rangeProof.alpha_hat: ${rp.get_alpha_hat()}`)
 *   console.log(`blsctData.rangeProof.tau_x: ${rp.get_tau_x()}`)
 *   console.log(`tokenId: token=${cTxOut.getTokenId().getToken()}, subid=${cTxOut.getTokenId().getSubid()}`)
 *   console.log(`vector_predicate: ${cTxOut.getVectorPredicate()}`)
 * }
 * const ser = cTx.serialize()
 * const deser = CTx.deserialize(ser)
 * ser === deser.serialize() // true 
 */
export class CTx extends ManagedObj {
  constructor(obj: any) {
    super(obj, () => deleteCTx(obj))
  }

  /** Constructs a new `CTx` instance.
   * @param srcTxIns - An array of `TxIn` objects representing the transaction inputs.
   * @param srcTxOuts - An array of `TxOut` objects representing the transaction outputs.
   * @returns A new `CTx` instance.
   */
  static generate(
    txIns: TxIn[],
    txOuts: TxOut[]
  ): CTx {
    // create vector and add txIns to it
    const txInVec = createTxInVec()
    for (const txIn of txIns) {
      addToTxInVec(txInVec, txIn.value())
    }

    // create vector and add txOuts to it
    const txOutVec = createTxOutVec()
    for (const txOut of txOuts) {
      addToTxOutVec(txOutVec, txOut.value())
    }

    const rv = buildCTx(txInVec, txOutVec)

    // free the temporary vectors
    deleteTxInVec(txInVec)
    deleteTxOutVec(txOutVec)

    if (rv.result === BLSCT_IN_AMOUNT_ERROR) {
      const msg = `Failed to build transaction. txIns[${rv.in_amount_err_index}] has an invalid amount`
      freeObj(rv)
      throw new Error(msg)
    }
    if (rv.result === BLSCT_OUT_AMOUNT_ERROR) {
      const msg = `Failed to build transaciton. tx_outs[${rv.out_amount_err_index}] has an invalid amount`
      freeObj(rv)
      throw new Error(msg)
    }

    if (rv.result !== 0) {
      const msg = `building tx failed. Error code = ${rv.result}`
      freeObj(rv)
      throw new Error(msg)
    }

    const obj = rv.ctx
    freeObj(rv)
    return CTx.fromObjAndSize(obj, 0)
  }

  override value(): any {
    return castToUint8_tPtr(this.obj)
  }

  /** Returns the transaction ID of this confidential transaction.
   * @returns The transaction ID.
   */
  getCTxId(): CTxId {
    const txIdHex = getCTxId(this.value())
    return CTxId.deserialize(txIdHex)
  }

  /** Returns the number of transaction inputs in this confidential transaction.
   * @returns The number of transaction inputs.
   */
  getCTxIns(): CTxIns {
    return new CTxIns(getCTxIns(this.value()))
  }

  /** Returns the number of transaction outputs in this confidential transaction.
   * @returns The number of transaction outputs.
   */ 
  getCTxOuts(): CTxOuts {
    return new CTxOuts(getCTxOuts(this.value()))
  }

  override serialize(): string {
    const buf = castToUint8_tPtr(this.value())
    return toHex(buf, this.size())
  }

  /** Deserializes a hexadecimal string into a `CTx` instance.
   * @param hex - The hexadecimal string to deserialize.
   * @returns A new `CTx` instance.
   */
  static deserialize(hex: string): CTx {
    if (hex.length % 2 !== 0) {
      hex = `0${hex}`
    }
    const objSize = hex.length / 2
    const obj = hexToMallocedBuf(hex)

    return CTx.fromObjAndSize(obj, objSize)
  }
}

File: ffi/rust/src/ctx.rs
use crate::{
  blsct_obj::{BlsctObj, self},
  blsct_serde::BlsctSerde, 
  ctx_id::CTxId,
  ctx_ins::CTxIns,
  ctx_outs::CTxOuts,
  ffi::{
    add_to_tx_in_vec,
    add_to_tx_out_vec,
    BlsctRetVal,
    BlsctCTx,
    BlsctCTxId,
    BLSCT_IN_AMOUNT_ERROR,
    BLSCT_OUT_AMOUNT_ERROR,
    build_ctx,
    create_tx_in_vec,
    create_tx_out_vec,
    delete_ctx,
    delete_tx_in_vec,
    delete_tx_out_vec,
    deserialize_ctx,
    deserialize_ctx_id,
    free_obj,
    get_ctx_id,
    get_ctx_ins,
    get_ctx_outs,
    serialize_ctx,
  },
  macros::{
    impl_clone,
    impl_display,
  },
  tx_in::TxIn,
  tx_out::TxOut,
};
use serde::{Deserialize, Serialize};
use std::{
  ffi::{
    c_char,
    c_void,
  },
  fmt,
  ptr::NonNull,
};

#[derive(Debug, PartialEq, Eq)]
pub enum Error {
  FailedToAllocateMemory,
  InAmountError(usize),
  OutAmountError(usize),
  FailedToBuildCTx(u8),
}

impl std::error::Error for Error {}

impl fmt::Display for Error {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    match self {
      Error::FailedToAllocateMemory =>
        write!(f, "Failed to allocate memory for CTx"),
      Error::InAmountError(index) =>
        write!(f, "Invalid in-amount found at {index}"),
      Error::OutAmountError(index) =>
        write!(f, "Invalid out-amount found at {index}"),
      Error::FailedToBuildCTx(e) =>
        write!(f, "Failed to build CTx: {e}"),
    }
  }
}

#[derive(Debug, Deserialize, Serialize, Eq)]
pub struct CTx {
  obj: BlsctObj<CTx, BlsctCTx>,
}

impl_display!(CTx);
impl_clone!(CTx);

impl CTx {
  pub fn new(
    tx_ins: &Vec<TxIn>,
    tx_outs: &Vec<TxOut>
  ) -> Result<Self, Error> {
    unsafe {
      let vp_tx_ins = create_tx_in_vec();
      let vp_tx_outs = create_tx_out_vec();

      for tx_in in tx_ins {
        add_to_tx_in_vec(vp_tx_ins, tx_in.value());
      }
      for tx_out in tx_outs {
        add_to_tx_out_vec(vp_tx_outs, tx_out.value());
      }
      let rv = build_ctx(vp_tx_ins, vp_tx_outs);
      if rv.is_null() {
        delete_tx_in_vec(vp_tx_ins);
        delete_tx_out_vec(vp_tx_outs);
        return Err(Error::FailedToAllocateMemory);
      }

      let clean_up = || {
        delete_tx_in_vec(vp_tx_ins);
        delete_tx_out_vec(vp_tx_outs);
        free_obj(rv as *mut c_void);
      };

      if (*rv).result == 0 {
        let vp_ctx = NonNull::<u8>::new((*rv).ctx as *mut u8).unwrap();
        let obj = BlsctObj::<CTx, BlsctCTx>::new_with_deallocator(vp_ctx, 0, Some(delete_ctx)); // size will not be used

        clean_up();
        Ok(obj.into())

      } else {
        let e = {
          match (*rv).result {
            BLSCT_IN_AMOUNT_ERROR => {
              let index = (*rv).in_amount_err_index;
              Error::InAmountError(index)
            },
            BLSCT_OUT_AMOUNT_ERROR => {
              let index = (*rv).out_amount_err_index;
              Error::OutAmountError(index)
            },
            err_code => {
              Error::FailedToBuildCTx(err_code)
            }
          }
        };
        clean_up();
        Err(e)
      } 
    }
  }

  pub fn get_ctx_id<'a>(&self) -> Result<CTxId, blsct_obj::Error<'a>> {
    let rv = unsafe { 
      let c_str_hex = get_ctx_id(self.value());
      deserialize_ctx_id(c_str_hex)
    };
    let obj = BlsctObj::<CTxId, BlsctCTxId>::from_retval(rv)?;
    Ok(obj.into())
  }

  pub fn get_ctx_ins(&self) -> CTxIns {
    let obj = unsafe { get_ctx_ins(self.value()) };
    obj.into()
  }

  pub fn get_ctx_outs(&self) -> CTxOuts {
    let obj = unsafe { get_ctx_outs(self.value()) };
    obj.into()
  }

  // not using impl_void_ptr_value!() to return *mut c_void
  // to avoid const_cast
  pub fn value(&self) -> *mut c_void {
    self.obj.as_ptr() as *mut c_void
  }
}

impl BlsctSerde for CTx {
  unsafe fn serialize(ptr: *const u8, _: usize) -> *const i8 {
    serialize_ctx(ptr as *mut c_void)
  }

  unsafe fn deserialize(hex: *const c_char) -> *mut BlsctRetVal {
    deserialize_ctx(hex)
  }
}

impl From<BlsctObj<CTx, BlsctCTx>> for CTx {
  fn from(obj: BlsctObj<CTx, BlsctCTx>) -> CTx {
    CTx { obj }
  }
}

impl PartialEq for CTx {
  fn eq(&self, other: &Self) -> bool {
    self.obj == other.obj
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::{
    ffi::{
      get_ctx_ins_size,
      get_ctx_outs_size,
    },
    initializer::init,
    test_util::gen_ctx,
  };

  #[test]
  fn test_get_ctx_id() {
    init();
    let ctx = gen_ctx();
    let _ = ctx.get_ctx_id();
  }

  #[test]
  fn test_get_ctx_ins() {
    init();
    let ctx = gen_ctx();
    let ctx_ins = ctx.get_ctx_ins();
    let ctx_ins_size = unsafe { get_ctx_ins_size(ctx_ins.value()) };
    assert_eq!(ctx_ins_size, 1);
  }

  #[test]
  fn test_get_ctx_outs() {
    init();
    let ctx = gen_ctx();
    let ctx_outs = ctx.get_ctx_outs();
    let ctx_outs_size = unsafe { get_ctx_outs_size(ctx_outs.value()) };
    assert_eq!(ctx_outs_size, 3);
  }

  #[test]
  fn test_deser() {
    init();
    let a = gen_ctx();
    let hex = bincode::serialize(&a).unwrap();
    let b = bincode::deserialize::<CTx>(&hex).unwrap();
    assert_eq!(a, b);
  }
}

